pkg/api/v1/defaults.go

```diff
@@ -165,6 +185,16 @@ func SetDefaults_NetworkInterface(obj *VirtualMachineInstance) {
 	}
 }
 
+func setDefaults_IOThreads(obj *VirtualMachineInstance) {
+	ioThreads := obj.Spec.Domain.IOThreads
+	if ioThreads != nil {
+		if ioThreads.ThreadCount == nil {
+			numIOThreads := uint(1)
```

Comment:
I think 1 is a good default because it can be problematic to calculate correct number of IO threads(you must have a lot of cluster information, that currently we do not have, and also if you requested CPU 200m how you will calculate expected number of IO threads), on the other side, 1 IO thread alread give to QEMU possibility to run all IO outside of the main lock.

Reply:
yes.
It's not a magic value which is best for all workloads but it helps the most in vast majority of cases where I/O is important. The only justification for more than 1 I've seen so far was in case of 2 separate disks going to different storage, reflected in the guest configuration, e.g. oracle db with separate writers for each

Reply:
CPU 200m still sits on 1 pCPU. 

On the other hand, when resources are bounded and limited, I highly doubt that we will achieve any acceleration. Separating the iothread is useful when there are enough free cycles, hence enough free pcpus. (IMHO) Therefore, I doubt we need to enable it when we don't have the resources, as this thread will just become a bottleneck.

I'm all for just making the right decision on behalf of the users, instead of letting them set the number. IOThread per disk is the best option I know. Regardless if QEMU is writing to different storage or not, afaik. This is about an i/o per disk being offloaded to a separate thread and it's best to have as many of these threads as there are disks, bounded by cpu of course. We already have bounded cpu resources in the pod, so I would just set the number of iothreads to the number of disks.

For cpu pinning, we will need to pin these threads to the already allocated set of dedicated cpus and pack the threads to the cpus. So if we have 6 disks and 4 cpus, 2 cpus with have 2 threads each and other 2 two only one.

Reply:
> On the other hand, when resources are bounded and limited, I highly doubt that we will achieve any acceleration. Separating the iothread is useful when there are enough free cycles, hence enough free pcpus. (IMHO) Therefore, I doubt we need to enable it when we don't have the resources, as this thread will just become a bottleneck.

to some degree, yes. That's why it shouldnt' be a default (not all workloads are I/O intensive) to not waste additional CPU cycles. It is slower actually, as the I/O tasks are needed to be handled by a separate thread instead. It only starts to matter when the I/O is heavy.

> I'm all for just making the right decision on behalf of the users, instead of letting them set the number. IOThread per disk is the best option I know. 

how would you account for that extra "waste" if they happen not to be highly used?

> Regardless if QEMU is writing to different storage or not, afaik. This is about an i/o per disk being offloaded to a separate thread and it's best to have as many of these threads as there are disks, bounded by cpu of course. We already have bounded cpu resources in the pod, so I would just set the number of iothreads to the number of disks.

that's not what the scale team results were. They were indicating "1" is the best even with multiple disks

> For cpu pinning, we will need to pin these threads to the already allocated set of dedicated cpus and pack the threads to the cpus. So if we have 6 disks and 4 cpus, 2 cpus with have 2 threads each and other 2 two only one.

ideally we should pin it to the numa nodes handling the actual physical access. I dobn't think we have the visibility right now though

Reply:
I forgot to mention that in oVirt we recently enabled it to 1 iothread by default for all server VMs (and immediately got hit by bugs specific to iothread,  but that's a separate story...). Unfortunately we do not have any experience/numbers to understand if it helped or rather hurt